solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java

```diff
@@ -172,19 +174,9 @@ public SlotAcc createSlotAcc(FacetContext fcontext, int numDocs, int numSlots) t
 
     if (fieldsArg == null && fcontext.isShard()) { // shard requests normally only ask for fl=id,score
       returnFields = new SolrReturnFields();
-    } else if (fieldsArg == null) {
-      returnFields = rb.rsp.getReturnFields();
-    } else if (fieldsArg instanceof String) {
-      returnFields = new SolrReturnFields((String)fieldsArg, fcontext.req);
     } else {
-      String[] fl = null;
-      if (fieldsArg instanceof List) {
-        List lst = (List)fieldsArg;
-        fl = (String[])lst.toArray(new String[lst.size()]);
-      } else {
-        fl = (String[])fieldsArg;
-      }
-      returnFields = new SolrReturnFields(fl, fcontext.req);
+      returnFields = rb.rsp.getReturnFields();
+      returnFields = buildReturnFields(fcontext.req, rb.rsp);
```

Comment:
not sure why this is setting the local var twice?

Reply:
huh - think I must have missed some code during a refactor.  I'll take a closer look and tweak it.

Reply:
I've sorted that out now (as a new commit I'll merge into the original later)